### DISCUSSION TITLE
Shared logic across AsyncNotifier(s)

### DISCUSSION THREAD

**Discussion by chimon2000** [rrousselGit/riverpod#2057] [Q&A]
I have an `AutoDisposeAsyncNotifier` and `AsyncNotifier` that I am trying to share logic between.  I have created a mixin for this that accesses the ref (defined as `AsyncNotifierProviderRef` in the mixin), but when defining ref I am given the following error trying to use it on the `AutoDisposeAsyncNotifier`:

```
'StoreMixin.ref' ('AsyncNotifierProviderRef<T> Function()') isn't a valid override of 'BuildlessAutoDisposeAsyncNotifier.ref' ('AutoDisposeAsyncNotifierProviderRef<T> Function()').
```

Here's an abridged version of the class:

```dart

mixin StoreMixin<State> {
  Future<State> update(
    FutureOr<State> Function(State) cb, {
    FutureOr<State> Function(Object err, StackTrace stackTrace)? onError,
  });
  AsyncNotifierProviderRef<State> get ref;
}
```

Wondering how I can resolve this issue?

**Comment by rrousselGit** [comment]
There appears to be a type incompatibility in the ref. Your mixin is using a non-autoDispose ref, but your notifier is using an autoDispose one.

Not sure if you can get around that.
